# Elikem Yaw Hamenoo

In [1]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer

In [2]:
df= pd.read_csv(r'C:\Users\AIX\Downloads\Life Expectancy Data.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

In [4]:
df.describe()

Year  Life expectancy   Adult Mortality  infant deaths  \
count  2938.000000       2928.000000      2928.000000    2938.000000   
mean   2007.518720         69.224932       164.796448      30.303948   
std       4.613841          9.523867       124.292079     117.926501   
min    2000.000000         36.300000         1.000000       0.000000   
25%    2004.000000         63.100000        74.000000       0.000000   
50%    2008.000000         72.100000       144.000000       3.000000   
75%    2012.000000         75.700000       228.000000      22.000000   
max    2015.000000         89.000000       723.000000    1800.000000   

           Alcohol  percentage expenditure  Hepatitis B       Measles   \
count  2744.000000             2938.000000  2385.000000    2938.000000   
mean      4.602861              738.251295    80.940461    2419.592240   
std       4.052413             1987.914858    25.070016   11467.272489   
min       0.010000                0.000000     1.000000       0.000000   
25%       0.877500                4.685343    77.000000       0.000000   
50%       3.755000               64.912906    92.000000      17.000000   
75%       7.702500              441.534144    97.000000     360.250000   
max      17.870000            19479.911610    99.000000  212183.000000   

              BMI   under-five deaths         Polio  Total expenditure  \
count  2904.000000         2938.000000  2919.000000         2712.00000   
mean     38.321247           42.035739    82.550188            5.93819   
std      20.044034          160.445548    23.428046            2.49832   
min       1.000000            0.000000     3.000000            0.37000   
25%      19.300000            0.000000    78.000000            4.26000   
50%      43.500000            4.000000    93.000000            5.75500   
75%      56.200000           28.000000    97.000000            7.49250   
max      87.300000         2500.000000    99.000000           17.60000   

       Diphtheria      HIV/AIDS            GDP    Population  \
count  2919.000000  2938.000000    2490.000000  2.286000e+03   
mean     82.324084     1.742103    7483.158469  1.275338e+07   
std      23.716912     5.077785   14270.169342  6.101210e+07   
min       2.000000     0.100000       1.681350  3.400000e+01   
25%      78.000000     0.100000     463.935626  1.957932e+05   
50%      93.000000     0.100000    1766.947595  1.386542e+06   
75%      97.000000     0.800000    5910.806335  7.420359e+06   
max      99.000000    50.600000  119172.741800  1.293859e+09   

        thinness  1-19 years   thinness 5-9 years  \
count            2904.000000          2904.000000   
mean                4.839704             4.870317   
std                 4.420195             4.508882   
min                 0.100000             0.100000   
25%                 1.600000             1.500000   
50%                 3.300000             3.300000   
75%                 7.200000             7.200000   
max                27.700000            28.600000   

       Income composition of resources    Schooling  
count                      2771.000000  2775.000000  
mean                          0.627551    11.992793  
std                           0.210904     3.358920  
min                           0.000000     0.000000  
25%                           0.493000    10.100000  
50%                           0.677000    12.300000  
75%                           0.779000    14.300000  
max                           0.948000    20.700000

In [5]:
df

Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01               71.279624         65.0      1154   
1                64     0.01               73.523582         62.0       492   
2                66     0.01               73.219243         64.0       430   
3                69     0.01               78.184215         67.0      2787   
4                71     0.01                7.097109         68.0      3013   
...             ...      ...                     ...          ...       ...   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
0     ...    6.0               8.16         65.0        0.1  584.259210   
1     ...   58.0               8.18         62.0        0.1  612.696514   
2     ...   62.0               8.13         64.0        0.1  631.744976   
3     ...   67.0               8.52         67.0        0.1  669.959000   
4     ...   68.0               7.87         68.0        0.1   63.537231   
...   ...    ...                ...          ...        ...         ...   
2933  ...   67.0               7.13         65.0       33.6  454.366654   
2934  ...    7.0               6.52         68.0       36.7  453.351155   
2935  ...   73.0               6.53         71.0       39.8   57.348340   
2936  ...   76.0               6.16         75.0       42.1  548.587312   
2937  ...   78.0               7.10         78.0       43.5  547.358878   

      Population   thinness  1-19 years   thinness 5-9 years  \
0     33736494.0                   17.2                 17.3   
1       327582.0                   17.5                 17.5   
2     31731688.0                   17.7                 17.7   
3      3696958.0                   17.9                 18.0   
4      2978599.0                   18.2                 18.2   
...          ...                    ...                  ...   
2933  12777511.0                    9.4                  9.4   
2934  12633897.0                    9.8                  9.9   
2935    125525.0                    1.2                  1.3   
2936  12366165.0                    1.6                  1.7   
2937  12222251.0                   11.0                 11.2   

      Income composition of resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                               0.470        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427      

# Data Preprocessing

In [6]:
#Checking for missing values

missing_values = df.isnull().sum()
print(missing_values)

Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64


In [7]:

num_imputer = SimpleImputer(strategy='mean')


num_columns = df.select_dtypes(include=[np.number]).columns


df[num_columns] = num_imputer.fit_transform(df[num_columns])


cat_imputer = SimpleImputer(strategy='most_frequent')


cat_columns = df.select_dtypes(include=[object]).columns


df[cat_columns] = cat_imputer.fit_transform(df[cat_columns])

In [8]:
#crosschecking for missing values
missing_values = df.isnull().sum()
print(missing_values)

Country                            0
Year                               0
Status                             0
Life expectancy                    0
Adult Mortality                    0
infant deaths                      0
Alcohol                            0
percentage expenditure             0
Hepatitis B                        0
Measles                            0
 BMI                               0
under-five deaths                  0
Polio                              0
Total expenditure                  0
Diphtheria                         0
 HIV/AIDS                          0
GDP                                0
Population                         0
 thinness  1-19 years              0
 thinness 5-9 years                0
Income composition of resources    0
Schooling                          0
dtype: int64


In [9]:
#checking for duplicates in the data

duplicates = df.duplicated().sum()
print(f'Duplicates: {duplicates}')
if duplicates > 0:
    df = df.drop_duplicates()

Duplicates: 0


In [10]:
#handling outliers in the data :

for col in num_columns:
    df[col] = df[col].clip(lower=df[col].quantile(0.01), upper=df[col].quantile(0.99))

In [11]:
#normalising the data 
from sklearn.preprocessing import StandardScaler

In [12]:
scaler = StandardScaler()
df[num_columns] = scaler.fit_transform(df[num_columns])

# Questions 

1.	Should a country having a lower life expectancy value (<65) increase its healthcare expenditure in order to improve its average lifespan?

In [13]:
low_life_expectancy = df[df['Life expectancy '] < 65]

In [14]:
correlation = low_life_expectancy['Total expenditure'].corr(low_life_expectancy['Life expectancy '])
print(f'Correlation: {correlation}')

Correlation: 0.21279862594286394


Yes, increasing healthcare expenditure may help improve life expectancy, but the impact is likely to be limited since the correlation coefficient of 0.2129 shows a weak positive relationship between healthcare expenditure and life expectancy.

2. How does Infant and Adult mortality rates affect life expectancy?

In [15]:
correlation_infant = df['infant deaths'].corr(df['Life expectancy '])
correlation_adult = df['Adult Mortality'].corr(df['Life expectancy '])

In [16]:
print(f'Correlation between infant deaths and Life expectancy: {correlation_infant}')
print(f'Correlation between Adult Mortality and Life expectancy: {correlation_adult}')

Correlation between infant deaths and Life expectancy: -0.30293963655460293
Correlation between Adult Mortality and Life expectancy: -0.696872133469117


-Infant Deaths: The correlation of -0.303 shows a moderate negative relationship, meaning higher infant deaths are associated with lower life expectancy.

-Adult Mortality: The correlation of -0.697 shows a strong negative relationship, meaning higher adult mortality significantly lowers life expectancy.

3.	Does Life Expectancy have positive or negative correlation with eating habits, lifestyle, exercise, smoking, drinking alcohol etc.

In [17]:
lifestyle_factors = ['Alcohol', ' BMI ']
correlations = {}

for factor in lifestyle_factors:
    correlation = df[factor].corr(df['Life expectancy '])
    correlations[factor] = correlation
    print(f'Correlation between {factor.strip()} and Life expectancy: {correlation}')

Correlation between Alcohol and Life expectancy: 0.3943112094955718
Correlation between BMI and Life expectancy: 0.5632480555590098


-Alcohol Consumption: The correlation of 0.394 shows a moderate positive relationship, meaning higher alcohol consumption is associated with higher life expectancy.

Also,

-BMI: The correlation of 0.563 shows a strong positive relationship, meaning higher BMI is associated with higher life expectancy. 

4.	What is the impact of schooling on the lifespan of humans?

In [18]:
correlation_schooling = df['Schooling'].corr(df['Life expectancy '])
print(f'Correlation between Schooling and Life Expectancy: {correlation_schooling}')

Correlation between Schooling and Life Expectancy: 0.732463026689371


The correlation of 0.732 shows a strong positive relationship between schooling on the lifespan of human. This means that higher levels of schooling are associated with higher life expectancy.

5.	Does Life Expectancy have a positive or negative relationship with drinking alcohol?

In [19]:
correlation_alcohol = df['Alcohol'].corr(df['Life expectancy '])
print(f'Correlation between Alcohol and Life Expectancy: {correlation_alcohol}')

Correlation between Alcohol and Life Expectancy: 0.3943112094955718


The correlation of 0.394 shows a moderate positive relationship, meaning higher alcohol consumption is associated with higher life expectancy.

6.	Do densely populated countries tend to have lower life expectancy?

In [20]:
correlation_population = df['Population'].corr(df['Life expectancy '])
print(f'Correlation between Population and Life Expectancy: {correlation_population}')

Correlation between Population and Life Expectancy: -0.014444651053079625


The correlation of -0.014 shows a very weak negative relationship, meaning there is almost no relationship between population size and life expectancy.

This means that densely populated countries do not tend to have significantly lower life expectancy.

7.	What is the impact of Immunization coverage on life Expectancy?

In [21]:
correlation_hepatitis_b = df['Hepatitis B'].corr(df['Life expectancy '])
print(f'Correlation between Hepatitis B and Life expectancy: {correlation_hepatitis_b}')

Correlation between Hepatitis B and Life expectancy: 0.2053382716994413


In [22]:
correlation_polio = df['Polio'].corr(df['Life expectancy '])
print(f'Correlation between Polio and Life expectancy: {correlation_polio}')

Correlation between Polio and Life expectancy: 0.4626908319145177


In [23]:
correlation_diphtheria = df['Diphtheria '].corr(df['Life expectancy '])
print(f'Correlation between Diphtheria and Life expectancy: {correlation_diphtheria}')

Correlation between Diphtheria and Life expectancy: 0.47652808230669663


Immunization coverage positively impacts life expectancy. While Hepatitis B shows a weak positive correlation, Polio and Diphtheria show moderate positive correlations with life expectancy. Increasing immunization coverage for these diseases can contribute to improving life expectancy.

8.	What are the predicting variables actually affecting life expectancy?

In [24]:
numeric_df = df.drop(columns=['Country', 'Status'])

In [25]:
correlation_matrix = numeric_df.corr()
correlation_life_expectancy = correlation_matrix['Life expectancy '].sort_values(ascending=False)
print("Correlation with Life Expectancy:")
print(correlation_life_expectancy)

Correlation with Life Expectancy:
Life expectancy                    1.000000
Schooling                          0.732463
Income composition of resources    0.693659
 BMI                               0.563248
Diphtheria                         0.476528
Polio                              0.462691
GDP                                0.448633
percentage expenditure             0.404050
Alcohol                            0.394311
Total expenditure                  0.212799
Hepatitis B                        0.205338
Year                               0.168722
Population                        -0.014445
Measles                           -0.183031
infant deaths                     -0.302940
under-five deaths                 -0.325339
 thinness 5-9 years               -0.487264
 thinness  1-19 years             -0.491692
 HIV/AIDS                         -0.591468
Adult Mortality                   -0.696872
Name: Life expectancy , dtype: float64


Based on the criterion of (|c| > 0.5), the significant predictors of life expectancy are:

-Positive Predictors: Schooling, income composition of resources, and BMI.

-Negative Predictors: Adult mortality, HIV/AIDS prevalence, and rates of thinness among children (ages 1-19 and 5-9 years).


9.	Train and test at least 3 regression models and pick the best performing model to predict life expectancy

In [26]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from sklearn.impute import SimpleImputer

In [27]:
features = ['Schooling', 'Income composition of resources', ' BMI ', 'Adult Mortality', ' HIV/AIDS', ' thinness  1-19 years', ' thinness 5-9 years']
X = df[features]
y = df['Life expectancy ']


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [28]:
# RMSE: Root Mean Squared Error
# MAE: Mean Absolute Error

In [29]:
#Linear Regression

lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)
r2_lr = r2_score(y_test, y_pred_lr)
rmse_lr = mean_squared_error(y_test, y_pred_lr, squared=False)
mae_lr = mean_absolute_error(y_test, y_pred_lr)
print(f"Linear Regression - R2 Score: {r2_lr:.4f}, RMSE: {rmse_lr:.4f}, MAE: {mae_lr:.4f}")

Linear Regression - R2 Score: 0.7916, RMSE: 0.4485, MAE: 0.3300


In [30]:
#Polynomial Regression

poly = PolynomialFeatures(degree=3)
X_poly_train = poly.fit_transform(X_train)
X_poly_test = poly.transform(X_test)
pr = LinearRegression()
pr.fit(X_poly_train, y_train)
y_pred_pr = pr.predict(X_poly_test)
r2_pr = r2_score(y_test, y_pred_pr)
rmse_pr = mean_squared_error(y_test, y_pred_pr, squared=False)
mae_pr = mean_absolute_error(y_test, y_pred_pr)
print(f"Polynomial Regression - R2 Score: {r2_pr:.4f}, RMSE: {rmse_pr:.4f}, MAE: {mae_pr:.4f}")

Polynomial Regression - R2 Score: 0.8550, RMSE: 0.3740, MAE: 0.2082


In [31]:
#Decision Tree Regressor

dt = DecisionTreeRegressor(random_state=42)
dt.fit(X_train, y_train)
y_pred_dt = dt.predict(X_test)
r2_dt = r2_score(y_test, y_pred_dt)
rmse_dt = mean_squared_error(y_test, y_pred_dt, squared=False)
mae_dt = mean_absolute_error(y_test, y_pred_dt)
print(f"Decision Tree Regressor - R2 Score: {r2_dt:.4f}, RMSE: {rmse_dt:.4f}, MAE: {mae_dt:.4f}")

Decision Tree Regressor - R2 Score: 0.9299, RMSE: 0.2601, MAE: 0.1536


In [32]:
if r2_lr > r2_pr and r2_lr > r2_dt:
    best_model = "Linear Regression"
    best_r2 = r2_lr
    best_rmse = rmse_lr
    best_mae = mae_lr
elif r2_pr > r2_lr and r2_pr > r2_dt:
    best_model = "Polynomial Regression"
    best_r2 = r2_pr
    best_rmse = rmse_pr
    best_mae = mae_pr
else:
    best_model = "Decision Tree Regressor"
    best_r2 = r2_dt
    best_rmse = rmse_dt
    best_mae = mae_dt

In [33]:
print(f"Best performing model: {best_model} with R2 Score: {best_r2:.4f}, Root Mean Squared Error: {best_rmse:.4f},  Mean Absolute Error: {best_mae:.4f}")

Best performing model: Decision Tree Regressor with R2 Score: 0.9299, Root Mean Squared Error: 0.2601,  Mean Absolute Error: 0.1536
